### ISSUE TITLE
I cannot get the bindgenhello example to work

### ISSUE THREAD

**Issue by andrewdavidmackenzie** [andrewdavidmackenzie/wasm-book#1]
I've tried coding the 'bindgenhello' example from the book.
It can be found in the 'bindgenhello' sub-folder of this repo:
https://github.com/andrewdavidmackenzie/wasm-book/tree/master/bindgenhello 

but I cannot get it working - I get a 404 loading the wasm.

I've added a Makefile to help build and run it. Using it to build and run the 'bindgenhello' example I get this output:

```
> make
npm run build-debug

> @ build-debug /Users/andrew/workspace/wasm-book/bindgenhello
> cargo +nightly build --target wasm32-unknown-unknown && wasm-bindgen target/wasm32-unknown-unknown/debug/bindgenhello.wasm --out-dir .

    Finished dev [unoptimized + debuginfo] target(s) in 0.22s
python3 -m http.server
Serving HTTP on 0.0.0.0 port 8000 (http://0.0.0.0:8000/) ...
127.0.0.1 - - [31/Dec/2018 19:28:10] "GET / HTTP/1.1" 200 -
127.0.0.1 - - [31/Dec/2018 19:28:10] "GET /bootstrap.js HTTP/1.1" 200 -
127.0.0.1 - - [31/Dec/2018 19:28:10] "GET /index.js HTTP/1.1" 200 -
127.0.0.1 - - [31/Dec/2018 19:28:10] "GET /bindgenhello.js HTTP/1.1" 200 -
127.0.0.1 - - [31/Dec/2018 19:28:10] code 404, message File not found
127.0.0.1 - - [31/Dec/2018 19:28:10] "GET /bindgenhello_bg HTTP/1.1" 404 -
127.0.0.1 - - [31/Dec/2018 19:28:10] "GET /favicon.ico HTTP/1.1" 200 -
```

The Console on the web browser says:
```
GET http://0.0.0.0:8000/bindgenhello_bg net::ERR_ABORTED 404 (File not found)
bootstrap.js:2 Error importing `index.js`: TypeError: Failed to fetch dynamically imported module: http://0.0.0.0:8000/index.js
```

I have other wasm-bindgen examples running (from the wasm-bindgen repo /examples folder), so it's not critical - but in case it is some issue with the code on the book...

**Comment by autodidaddict** [comment]
Oddly enough, in the time since I wrote that original sample, browser support for that particular syntax has changed. I am updating the book code but the following should get you running on webpack.

New webpack.config.js:
```
const path = require('path');
const HtmlWebpackPlugin = require('html-webpack-plugin');
const webpack = require('webpack');

module.exports = {
    entry: './index.js',
    output: {
        path: path.resolve(__dirname, 'dist'),
        filename: 'index.js',
    },
    plugins: [
        new HtmlWebpackPlugin(),
        // Have this example work in Edge which doesn't ship `TextEncoder` or
        // `TextDecoder` at this time.
        new webpack.ProvidePlugin({
            TextDecoder: ['text-encoding', 'TextDecoder'],
            TextEncoder: ['text-encoding', 'TextEncoder']
        })
    ],
    mode: 'development'
};
```

new index.js:
```
const wasm = import('./bindgenhello');

wasm
    .then(h => h.hello("world!"))
    .catch(console.error);
```

If you manually run the `cargo build` command to produce a .wasm file, and then manually run the `wasm-bindgen` command, it'll produce the necessary JS file and then you'll be able to run `npm install` and `npm run serve`.

running python against an index.html will not currently work. 

Thanks for the bug report and I'll make sure that the chapter is updated to take all this into account.

**Comment by andrewdavidmackenzie** [comment]
`npm run serve` was failing until I manually installed "html-webpack-plugin" with:
> `npm i --save-dev html-webpack-plugin`

Then `npm run serve` was failing with this:
```
ERROR in ./bindgenhello.js
Module not found: Error: Can't resolve 'text-encoding' in '/Users/andrew/workspace/wasm-book/bindgenhello'
 @ ./bindgenhello.js 1:0-24 1:0-24
 @ ./index.js
 @ multi (webpack)-dev-server/client?http://localhost:8080 ./index.js
```

I manually installed 'text-encoding' with:
> `npm i --save-dev text-encoding`

And now it works! Opening http://localhost:8080 and the Alert pops up.

I am not sure why I had to install those dependencies manually and `npm install` didn't work. I don't know enough about node, npm an company to solve that one.

I suspected the package.json declared `devDependencies` map instead of `dependencies`, but trying that alternative didn't help.

Either way - I got it working. I modified my Makefile to do the steps (as I'm more proficient in that than npm and webpack...) in case it's useful. I will do a PR to my own repo with the changes and close this issue with it.
